### PR TITLE
Introduce Certified-Address-Book interface for Peer-Store

### PR DESCRIPTION
**Pull request by lla-dane**
Tracks #638 

This PR implements the `CertifiedAddressBook` interface (signed-peer-records) for `py-libp2p` peer-store.

### CODE REVIEW

**Comment by lla-dane** [comment]
## Current Progress:
#### `PeerRecord` Class: 
The core `PeerRecord` class is completed in `libp2p/peer/peer_record.py`. This class serves as the fundamental building block for managing cetified peer information and includes:

* `peer_id`: The peer's globally unique identifier (ID type)

* `addrs`: List of publicly reachable multiaddresses.

* `seq`: Strictly monotonically increasing sequence number for versioning.

### Key Methods:
* `domain`: Returns the domain string `"libp2p-peer-record"` for envelope validation

* `codec`: Returns binary codec prefix `(b"\x03\x01")` for type identification

* `to_protobuf`: Converts PeerRecord to protobuf representation

* `marshal_record`: Serializes record to bytes for envelope embedding

* `equal`: Compares two PeerRecord instances for equality
 
#### `IPeerRecord` Interface Definition in `abc.py`:
A comprehensive abstract interface (`IPeerRecord`) has been added to `libp2p/abc.py` that defines the contract for all PeerRecord implementations, ensuring consistency across the system.

#### Utility functions:
Several utility function have been implemented to support `PeerRecord` ecosystem:
* `unmarshal_record`: Deserializes PeerRecord from bytes.

* `timestamp_seq`: Generates strictly increasing sequence numbers with thread safety.

* `peer_record_from_peer_info`: Creates PeerRecord from existing PeerInfo objects

* `peer_record_from_protobuf`: Converts protobuf messages to PeerRecord instances.

* `addrs_from_protobuf`and `addrs_to_protobuf`: Handles address conversion.

There is an extra addition here regarding the use of `ProtoBuf` with the `PeerRecord`:
ProtocolBuffers (ProtoBuf) is a language-agnostic data serialization format developed by Google. Think of it as a compact, binary version of JSON or XML, but much smaller, faster, and more suitable for systems-level-messaging.

**Key Features:**
* Strongly typed and **Schema driven**
* Efficient for both network transmission and disk storage.
* Compatible across programming languages.
* Allows versioning (backward/forward compatibility)
This makes ideal for peer-to-peer systems like libp2p, where efficiency and compatibility are critical.

### What Protobuf does in this Context (PeerRecord)
In libp2p, `PeerRecord` is a standardized message format for representing metadata about a peer. 

The `.proto` file defines:

```proto
message PeerRecord {
    message AddressInfo {
        bytes multiaddr = 1;
    }
    bytes peer_id = 1;
    uint64 seq = 2;
    repeated AddressInfo addresses = 3;
}
```
Which encodes:
* `peer_id`: The peer's identity.
* `seq`: A monotonic timestamp (used for freshness comparison).
* `addresses`: A list of `multiaddr` objects where the peer can be reached.

This schema is compiled by `protoc` into languge-specific representations (in `peer_record_pb2.py` in Python, [peer.pb.go](https://github.com/libp2p/go-libp2p/blob/master/core/peer/pb/peer_record.pb.go) in Go), which applications use to **encode and decode PeerRecords** into raw bytes in a language neutral format.

### Role of `ProtoBuf` at runtime:
* **Serialization**: When a peer wants to send its metadata (eg. via `identify, discovery` or peer routing), it constructs a `PeerRecord` and uses the generated code to it into bytes.
* **Deserialization**: When another peer receives it, it uses the same schema to parse those bytes back into a usable object.

This is done through the generated classes like `pb2.PeerRecord`, where `to_protobuf` and `from_protobuf` logic in the code converts between domain objects and serialized form.

More importantly, because the schema is shared across implementations (eg, Go, Rust, Js, Python), everyone knows exactly how to interpret the bytes, regardless the language. 

Referances from go-libp2p:
* peer_record.proto: Used for the protobuf schema in the python side: https://github.com/libp2p/go-libp2p/blob/master/core/peer/pb/peer_record.proto
* PeerRecord interface: https://github.com/libp2p/go-libp2p/blob/master/core/peer/record.go

**Comment by lla-dane** [comment]
Next will start with the `Envelope` interface. Reference from go-lib2p: https://github.com/libp2p/go-libp2p/blob/master/core/record/envelope.go

@seetadev

**Comment by lla-dane** [comment]
`Envelope` wrapper class is completed with tests and docstrings. Created separate `envelope.proto` and `crypto.proto` schemas exactly same as that of go-libp2p for compatibility. 

Referance from go-libp2p: 
* `Envelope` wrapper interface: https://github.com/libp2p/go-libp2p/blob/master/core/record/envelope.go
* `envelope.proto` schema: https://github.com/libp2p/go-libp2p/blob/master/core/record/pb/envelope.proto
* `crypto.proto` schema: https://github.com/libp2p/go-libp2p/blob/master/core/crypto/pb/crypto.proto

Now remains, connecting the `PeerRecord` and `Envelope` class with a `CertifiedAddrBook` interface in `PeerStore`. 

### IMPORTANT
One more thing is that had to ignore the typechecks on the protobuf injected code objects, have to figure out how to make the protobuf generated code available to the linter. @seetadev 

Ignored typeschecks for protobuf injected code like this in many places:
```py
def to_protobuf(self) -> pb.PeerRecord:  # type: ignore[attr-defined, name-defined]
```

**Comment by lla-dane** [comment]
The internal working of the `CertifiedAddrBook` interface with all its surrounding utilties like `PeerRecord` class and `Envelope` class is completed. 
* The relevant tests to understand the working of these classes together and separately has been added along with detailed docstrings. 

* Open to suggestions for adding tests for more edge cases. 

* Improvement feedback on code design would also be helpful. 

* Added a few TODOs in the `peerstore.py`:
	```py
	TODO: Set up an async task for periodic peer-store cleanup for expired addresses and records.
	
	TODO: Make proper use of this function (maybe_delete_peer_record)
	
	TODO: Put up a limit on the number of records to be stored ?

	TODO: In case of overwriting a record, what should be do with the old addresses, do we overwrite them with the new addresses too ?
	``` 
	Few pointers on these would be helpful if we are to solve these in this PR only.
	
* Would also need pointers on how to set up the mechanism for transfering/receiving peer rocords, like from which interface we accept/send them (processing wont be much bigger issue), if we are to set this up in this PR only, otherwise could open up an issue for this.


> One more thing is that had to ignore the typechecks on the protobuf injected code objects, have to figure out how to make the protobuf generated code available to the linter. @seetadev 

Pointers on this would also be helpful. 


Other than these, ready for peer review on the internal working of the `CertifiedAddrBook`. @seetadev  @pacrob

**Comment by seetadev** [comment]
@lla-dane : Hi Abhinav. Great work indeed. Reviewing the code commits and the peer store additions. Appreciate your great effort.

CCing @Winter-Soren, @acul71, @mystical-prog, @guha-rahul  and @sumanjeet0012 for their feedback and pointers on the functionality enhancement. Indeed, this is needed both by gossipsub 1.1 upgrade effort as well as pub-sub improvements too.

**Comment by acul71** [comment]
@lla-dane @seetadev 
Discussion about Certified Address Book TODOs here:
TODOs of WIP: Certified-Address-Book interface for Peer-Store https://github.com/libp2p/py-libp2p/discussions/777

**Comment by lla-dane** [comment]
Thanks @acul71: for the detailed pointers. Will start reflecting on the suggestions.

**Comment by lla-dane** [comment]
@seetadev @acul71: Plese take a look at this commit a35225f. I have integrated the Signed peer-record transfer with the identify/identify-push protocols. It was done in a similar way n go-libp2p also. 
Updated the `identify-demo` example also to show the received peer-record from the sender. 
Added extra checks in the identify tests related to authenticity of the received peer-records. 
Also have solved a few todos as suggested by @acul71.
Please have a thorough look on the transfer mechanism and see if everything looks good.

**Comment by seetadev** [comment]
@lla-dane : Hi Abhinav. Thanks for the fantastic work on integrating the **Signed Peer Record** transfer into the `identify` and `identify-push` protocols — we are reviewing your commit [a35225f](https://github.com/libp2p/py-libp2p/commit/a35225f), and it’s clear you've covered a lot of ground here.

We especially appreciate the following:

* **Alignment with go-libp2p:** It's great to see that you've mirrored the implementation strategy from `go-libp2p`, making future interoperability and feature parity smoother across implementations.
* **Updates to the `identify-demo` example:** The demo makes it much easier to see the peer record transfer mechanism in action. This will be really helpful for other contributors and testers alike.
* **Authenticity checks in the `identify` tests:** These are crucial for maintaining correctness and trust in the record exchange — well done on proactively addressing this.
* **Resolution of outstanding TODOs:** Thanks for following up on the feedback from @acul71 and helping move the module toward completeness and maintainability.

We'll do a thorough review of the peer record transfer logic as requested. At a first glance, the transfer mechanism looks solid and integrates cleanly. We'll also test a few edge cases around invalid/malformed records to ensure the checks behave as expected.

This PR is a significant step toward supporting a **certified address book** in `py-libp2p`, and it's exciting to see it come together.

We'll follow up shortly with any comments or suggestions after deeper testing. Thanks again for your thoughtful and well-scoped implementation. 

CCing @acul71 and @Winter-Soren for their review, feedback and pointers too. This indeed will help not just gossipsub, pub-sub efforts but also NAT traversal initiatives in a fine manner.

File: libp2p/peer/peerstore.py [hunk 1]
```diff
@@ -165,6 +215,94 @@ def clear_metadata(self, peer_id: ID) -> None:
...
+            except PeerStoreError:
+                existing_addrs = set()
+        else:
+            existing_addrs = set()
+
+        new_addrs = set(record.addrs)
+        merged_addrs = list(existing_addrs.union(new_addrs))
+
+        self.peer_record_map[peer_id] = PeerRecordState(envelope, record.seq)
+        self.add_addrs(peer_id, merged_addrs, ttl)
```
Comment (lla-dane) [inline libp2p/peer/peerstore.py]: @seetadev @acul71: Please take a look at this part. 
In case when when we receive a signed-peer-record and the are existing addrs corresponding to that `peer_id` in the `peer_data_map`. So what is happening presently is we are merging the `existing_addrs` and `new_addrs`(which come with the peer-record) here: and adding the union of both in to the `peer_data_map`. 
```py
merged_addrs = list(existing_addrs.union(new_addrs))
```
So I am thinking we should overwrite the `existing_addrs` with `new_addrs` and make it so that there are only `new_addrs` in the `peer_data_map` for that `peer_id` as they come freshly signed and attested from the peer. What do you think about this?

File: examples/ping/ping.py [hunk 2]
```diff
@@ -59,6 +59,9 @@ async def run(port: int, destination: str) -> None:
     host = new_host(listen_addrs=[listen_addr])
 
     async with host.run(listen_addrs=[listen_addr]), trio.open_nursery() as nursery:
+        # Start the peer-store cleanup task
+        nursery.start_soon(host.get_peerstore().start_cleanup_task, 60)
```
Comment (lla-dane) [inline examples/ping/ping.py:63]: @seetadev: started an async Peerstore cleanup task in the ping-example. Should I do similar thing in other other examples also. 
Also have written a test on the cleanup task in `test_peerstore.py` file to verify the behaviour.

Comment (seetadev) [reply examples/ping/ping.py:63]: @lla-dane : Great work, Abhinav. Yes, please follow a similar workflow for other examples. This PR is coming along nicely. Appreciate your efforts and initiative.

**Comment by lla-dane** [comment]
@seetadev: I have added the peeriodic peer-store cleanup in all the examples as suggested. Now this PR could be reviewed further. All the general testing is done. You can check them out in the File changed section. 
So further feedback and improvement suggestions would be great.

**Comment by lla-dane** [comment]
Also suggest, other than `identify` protocol, with which protocols do I need to integrate the signed-peer-record transfer.

**Comment by seetadev** [comment]
@lla-dane : Thanks a lot for pushing this forward and for incorporating the periodic peer-store cleanup across all the examples — that's a great improvement.

I am reviewing the changes in the Files changed section. Looks clean and well-structured at the first glance. Will share feedback after detailed review.

The general testing you've done also gives good confidence in the stability of this integration. Great job making this consistent across the board — it really helps the examples become more production-aligned and avoids peer-store bloat in long-running sessions.

As for next steps, here are a couple of thoughts:

Review Suggestions
It might be useful to log or expose the cleanup interval as a tunable parameter in the examples. This could help folks experimenting with different peer churn scenarios.

If possible, you could add a small inline comment or docstring near the cleanup loop in each example, just to briefly describe its purpose for new readers.

On Signed Peer Records
You're absolutely right to consider expanding beyond identify when it comes to signed peer record (SPR) transfer.

In addition to identify, you may want to consider these protocols:

GossipSub (PubSub)

Peer records can be attached to control messages or even as metadata to help bootstrap peer scoring.

Particularly useful in large distributed applications (e.g. blockchain clients, chat apps).

@sumanjeet0012 can help here, especially in Pub sub example.


DHT (Kad)

When dialing or updating peer addresses in the DHT routing table, SPRs can provide a trusted mapping from peer ID to multiaddrs.

Could help with more secure peer discovery and address updates.

@sumanjeet0012 can again help here.


Relay (Circuit v2)

Peers establishing relayed connections could use SPRs to validate remote addresses or introduce themselves with additional metadata.

@Winter-Soren and @guha-rahul can help here.


Connection Gating or Peer Routing Extensions

If we extend py-libp2p with custom protocols later for reputation or trust scoring, SPRs could become part of those payloads as well.

@acul71 could help here.

Great job strengthening the py-libp2p stack and making the examples more robust and educational. Looking forward to reviewing this PR in more detail and iterating on any feedback.

**Comment by lla-dane** [comment]
> It might be useful to log or expose the cleanup interval as a tunable parameter in the examples. This could help folks experimenting with different peer churn scenarios.
If possible, you could add a small inline comment or docstring near the cleanup loop in each example, just to briefly describe its purpose for new readers.

Cleanup interval is tunable and docstrings single line comments are added in each example upon starting the peer-store cleanup task.

**Comment by seetadev** [comment]
> It might be useful to log or expose the cleanup interval as a tunable parameter in the examples. This could help folks experimenting with different peer churn scenarios.
> If possible, you could add a small inline comment or docstring near the cleanup loop in each example, just to briefly describe its purpose for new readers.

@lla-dane : Great, thank you for your quick response and for addressing the questions and feedback. Appreciate it.

We could update the examples separately in another PR, if you wish. This might make this PR easier to navigate for the new developers.

Please also add developer notes in detail along with a screencast for the signed-peer-record transfer in the identify example at the discussion page. Appreciate your efforts and initiative.

**Comment by lla-dane** [comment]
@seetadev: added the newsfragment.

**Comment by seetadev** [comment]
@lla-dane : Great, thank you for your quick response and for addressing the feedback points. Appreciate it.

Please also add developer notes in detail along with a screencast for the signed-peer-record transfer in the identify example at the discussion page. Also, open up the project tracking issue as discussed. We will update the examples separately in another PR, if you wish. This might make this PR easier to navigate for the new developers. Appreciate your efforts and initiative.

This PR is ready for final review and merge :)

**Comment by lla-dane** [comment]
Developer noted regarding how the signed-peer-record transfer has been posted to https://github.com/libp2p/py-libp2p/discussions/739#discussioncomment-13899845
